<a href="https://colab.research.google.com/github/lcbjrrr/quant/blob/master/Modelo_de_Leontief.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Ficha Técnica
Disciplina - Algebra linear e Processos Dinâmicos.

Conteúdo -

Aplicações de Álgebra Linear.

Modelos de Leontief.

Autores:

Anderson Borba (anderson.aborba@professores.ibmec.edu.br) e

Luiz Barboza (luiz.cbjunior@professores.ibmec.edu.br).

Endereço eletrônico: http://quant-ibmec.group/calculos/

# Aplicação de álgebra linear.
- Modelos econômicos para troca de Bens (Modelos de Leontief).
- Modelo ${\bf x}= C{\bf x}+ {\bf d}$.
- Vetor produção (Valores em unidades monetárias).
\begin{equation}
        {\bf x} = 
    \left[
    \begin{array}{c}
        x_1  \\
        x_2  \\
        \vdots\\
        x_n  \\
    \end{array}
    \right]
\end{equation}
- Vetor demanda.
\begin{equation}
        {\bf d} = 
    \left[
    \begin{array}{c}
        d_1  \\
        d_2  \\
        \vdots\\
        d_n  \\
    \end{array}
    \right]
\end{equation}
- Matriz de consumo.
\begin{equation}
    \left[
    \begin{array}{cccc}
        c_{11}& c_{12}& \cdots& c_{1n} \\
        c_{21}& c_{22}& \cdots& c_{2n} \\
        \vdots & \vdots&  &\vdots \\
        c_{n1}& c_{n2}& \cdots& c_{nn} \\
    \end{array}
    \right]
\end{equation}
- $C_{ij}$ representam as vendas de $i$ para $j$.
# Modelos de Leontief
  - ${\bf x}=C{\bf x}+{\bf d}$.
  - ${\bf x}-C{\bf x}={\bf d}$.
  - $(I-C){\bf x}={\bf d}$.
  
Resolvendo esse sistema linear teremos o vetor de produção ${\bf x}$ em valores de unidades monetárias.

## Exemplo
Seja ${\bf x}$ o vetor produção
\begin{equation}\nonumber
    \left[
    \begin{array}{c}
	    x_1 \\
	    x_2 \\
	    x_3\\
    \end{array}
    \right].
\end{equation}
Seja ${\bf d}$ o vetor demanda externa
\begin{equation}\nonumber
    \left[
    \begin{array}{c}
        d_1 \\
        d_2 \\
        d_3 \\
    \end{array}
    \right]=
    \left[
    \begin{array}{c}
        50 \\
        30 \\
        20 \\
    \end{array}
    \right].
\end{equation}
 - Índice 1 - Manufatura.
 - Índice 2 - Agricultura.
 - Índice 3 - Serviços.
  
E seja $C$ a matriz de consumo
\begin{equation}
    \left[
    \begin{array}{cccc|r}
       0.50   & 0.40 & 0.20\\
       0.20 & 0.30   & 0.10\\
       0.10 & 0.10 & 0.30\\
    \end{array}
    \right].
\end{equation}
Para saber o valor em unidades monetárias produzida vamos resolver $(I-C){\bf x}={\bf d}$. Portanto,
\begin{equation}\nonumber
    \left[
    \begin{array}{c}
        x_1 \\
        x_2 \\
        x_3 \\
    \end{array}
    \right]=
    \left[
    \begin{array}{c}
        225.92\\
        118.51 \\
        77.77\\
    \end{array}
    \right].
\end{equation} 

# Resolução do sistema linear
## Eliminação de Gauss

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.linalg as la

C = np.array([[0.5,0.4,0.2],[0.2,0.3,0.1],[0.1,0.1,0.3]])
print(C)

[[0.5 0.4 0.2]
 [0.2 0.3 0.1]
 [0.1 0.1 0.3]]


In [ ]:
I = np.identity(3, dtype=float)
print(I)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
A = I - C
print(A)

[[ 0.5 -0.4 -0.2]
 [-0.2  0.7 -0.1]
 [-0.1 -0.1  0.7]]


### Matriz extendida

In [ ]:
AE = np.array([[0.5,-0.4,-0.2,50.0],[-0.2,0.7,-0.1,30.0],[-0.1,-0.1,0.7,20.0]])
print(AE)

[[ 0.5 -0.4 -0.2 50. ]
 [-0.2  0.7 -0.1 30. ]
 [-0.1 -0.1  0.7 20. ]]


### Operações com linhas e matrizes elementares

Definindo a operação $L_2=>L_2-\frac{a_{21}}{a_{11}}L_1$, isto é, $L_2=>L_2+\frac{2}{5}L_1$

In [ ]:
E1 = np.array([[1,0,0],[2/5,1,0],[0,0,1]])
print(E1)

[[1.  0.  0. ]
 [0.4 1.  0. ]
 [0.  0.  1. ]]


In [ ]:
E1 @ AE

array([[ 0.5 , -0.4 , -0.2 , 50.  ],
       [ 0.  ,  0.54, -0.18, 50.  ],
       [-0.1 , -0.1 ,  0.7 , 20.  ]])

Definindo a operação $L_3=>L_3-\frac{a_{31}}{a_{11}}L_1$, isto é, $L_3=>L_3+\frac{1}{5}L_1$

In [ ]:
E2 = np.array([[1,0,0],[0,1,0],[1/5,0,1]])
print(E2)

[[1.  0.  0. ]
 [0.  1.  0. ]
 [0.2 0.  1. ]]


In [ ]:
E2 @ E1 @ AE

array([[ 0.5 , -0.4 , -0.2 , 50.  ],
       [ 0.  ,  0.54, -0.18, 50.  ],
       [ 0.  , -0.18,  0.66, 30.  ]])

Definindo a operação $L_3=>L_3-\frac{a_{32}}{a_{22}}L_2$, isto é, $L_3=>L_3+\frac{1}{3}L_2$

In [ ]:
E3 = np.array([[1,0,0],[0,1,0],[0,1/3,1]])
print(E3)

[[1.         0.         0.        ]
 [0.         1.         0.        ]
 [0.         0.33333333 1.        ]]


In [ ]:
U = E3 @ E2 @ E1 @ AE
print(U)

[[ 5.00000000e-01 -4.00000000e-01 -2.00000000e-01  5.00000000e+01]
 [ 0.00000000e+00  5.40000000e-01 -1.80000000e-01  5.00000000e+01]
 [ 1.38777878e-17 -5.55111512e-17  6.00000000e-01  4.66666667e+01]]


Solução do sistema linear (retrosubstituíção)

In [ ]:
x3 = U[2,3] / U[2,2]
x2 = (U[1,3] - U[1,2] * x3) / U[1,1]
x1 = (U[0,3] - U[0,2] * x3 - U[0,1] * x2) / U[0,0]
print(x1)
print(x2)
print(x3)

225.92592592592595
118.51851851851853
77.77777777777779


Solução do sistema linear $x=(x_1,x_2,x_3)=(225.9259, 118.5185, 77.7777)$.

### Solução usando a matriz inversa de (I-C)
Seja $A = (I-C)$ temos que encontrar $A^{-1}$ para calcula $x=A^{-1}d$.

In [ ]:
d = np.array([[50],[30],[20]])
print(d)

[[50]
 [30]
 [20]]


In [ ]:
AINV = np.linalg.inv(A)
x = AINV @ d
print(x)


[[225.92592593]
 [118.51851852]
 [ 77.77777778]]


In [ ]:
AINV

array([[2.96296296, 1.85185185, 1.11111111],
       [0.92592593, 2.03703704, 0.55555556],
       [0.55555556, 0.55555556, 1.66666667]])

Solução do sistema linear $x=(x_1,x_2,x_3)=(225.9259, 118.5185, 77.7777)$. 

# Referências
- Anton, Howard Rorres, Chris. Álgebra Linear com Aplicações. Editora Bookman - 2012.
- Carl D, Meyer, Matrix Analysis and Applied Linear Algebra, Editora SIAM - 2000  